cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/service/http/BodyConsumerAdapter.java

```diff
@@ -0,0 +1,195 @@
+              }
+            }
+          });
+        } finally {
+          ClassLoaders.setContextClassLoader(oldClassLoader);
+        }
+      } else {
+        transactional.execute(new TxRunnable() {
+          @Override
+          public void run(DatasetContext context) throws Exception {
+            delegate.onReceived(request.toByteBuffer());
+          }
+        });
+      }
+    } catch (Throwable t) {
+      onError(t, this.responder);
+    }
+  }
+
+  @Override
+  public void finished(HttpResponder responder) {
+    try {
+      transactional.execute(new TxRunnable() {
+        @Override
+        public void run(DatasetContext context) throws Exception {
+          delegate.onFinish(BodyConsumerAdapter.this.responder);
+        }
+      });
+      BodyConsumerAdapter.this.responder.execute();
+    } catch (Throwable t) {
+      onError(t, this.responder);
+    }
+  }
+
+  @Override
+  public void handleError(final Throwable cause) {
+    // When this method is called from netty-http, the response has already been sent, hence uses a no-op
+    // DelayedHttpServiceResponder for the onError call.
+    onError(cause, new DelayedHttpServiceResponder(responder) {
+      @Override
+      protected void doSend(int status, ByteBuffer content, String contentType,
+                            @Nullable Multimap<String, String> headers, boolean copy) {
+        // no-op
+      }
+
+      @Override
+      public void setTransactionFailureResponse(Throwable t) {
+        // no-op
+      }
+
+      @Override
+      public void execute() {
+        // no-op
+      }
+    });
+  }
+
+  /**
+   * Calls the {@link HttpContentConsumer#onError(HttpServiceResponder, Throwable)} method from a transaction.
+   */
+  private void onError(Throwable cause, final DelayedHttpServiceResponder responder) {
+    final Throwable failureCause = (cause instanceof WrappedException) ? cause.getCause() : cause;
+    try {
+      transactional.execute(new TxRunnable() {
+        @Override
+        public void run(DatasetContext context) throws Exception {
+          delegate.onError(responder, failureCause);
+        }
+      });
+    } catch (Throwable t) {
+      responder.setTransactionFailureResponse(t);
+    } finally {
+      responder.execute();
+    }
+  }
+
+  private Transactional createTransactional(final TransactionContext txContext,
+                                            final DatasetContext datasetContext) {
+    return new Transactional() {
+      @Override
+      public void execute(final TxRunnable runnable) {
+        try {
+          txContext.start();
+          try {
+            ClassLoader oldClassLoader = ClassLoaders.setContextClassLoader(contextClassLoader);
+            try {
+              runnable.run(datasetContext);
+            } finally {
+              ClassLoaders.setContextClassLoader(oldClassLoader);
+            }
+            txContext.finish();
+          } catch (Throwable t) {
```

Comment:
I think this is wrong. The txContext can throw TransactionConflictException or other exceptions (for example if it can't talk to transaction service). In that case it is not a user handler exception. I would think that in case we need a catch around the runnable.run(), which wraps any exception into a TxFailureException. For finish, we should not catch exceptions, because it only throws TransactionFailureException (or its subclass TransactionConflictException). Not sure why we need WrappedException?